README.md
# SimpleHttp

TODO: Write a gem description

## Installation

Add this line to your application's Gemfile:

    gem 'simple_http'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install simple_http

## Usage

TODO: Write usage instructions here

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## 0.1.0
===================
- Initial commit

spec/spec_helper.rb
unless ENV['CI']
  require 'simplecov'
  SimpleCov.start do
    add_filter 'spec'
  end
end

require 'rspec'
require 'rspec/autorun'
require 'webmock/rspec'
require 'ostruct'
require 'simple_http'

RSpec.configure do |config|
  config.mock_with :rspec
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end

def stub_delete(path='')
  stub_request(:delete, "https://example.com:443#{path}")
end

def stub_get(path='')
  stub_request(:get, "https://example.com:443#{path}")
end

def stub_post(path='')
  stub_request(:post, "https://example.com:443#{path}")
end

def stub_put(path='')
  stub_request(:put, "https://example.com:443#{path}")
end

spec/connection_spec.rb
require File.expand_path('../spec_helper', __FILE__)

describe SimpleHttp::Connection do

  subject do
    @conn = SimpleHttp::Connection.new('https://example.com')
  end

  context "with user options" do
    before do
      @options = {
        :headers => {
          'Accept'         => 'application/json',
          'User-Agent'     => "Simple HTTP gem #{SimpleHttp::Version}",
          'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3'
        },
        :ssl => {:verify => false},
        :max_redirects => 2
      }
      @conn = SimpleHttp::Connection.new('https://microsoft.com', @options)
    end

    it "overrides default options" do
      opts = SimpleHttp::Connection.default_options
      opts.keys.each do |key|
        expect(@conn.instance_variable_get(:"@#{key}")).to eq @options[key]
      end
    end
  end

  describe "#default_headers" do
    it "returns user_agent and response format" do
      expect(subject.default_headers).to eq ({
        "Accept"          => "application/json", 
        "User-Agent"      => "Simple HTTP gem #{SimpleHttp::Version}",
        'Accept-Encoding' => 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3'
      })
    end
  end

  describe "#scheme" do
    it "returns the http scheme" do
      expect(subject.scheme).to eq 'https'
    end
  end

  describe "#scheme" do
    context "scheme is unsupported" do
      it "raises an error" do
        expect { subject.scheme = 'ftp'}.to raise_error(SimpleHttp::Connection::UnsupportedSchemeError)
      end
    end

    context "scheme is http" do
      it "sets the scheme" do
        subject.scheme = 'http'
        expect(subject.scheme).to eq 'http'
      end
    end

    context "scheme is https" do
      it "sets the scheme" do
        subject.scheme = 'https'
        expect(subject.scheme).to eq 'https'
      end
    end
  end

  describe "#host" do
    it "returns the host server" do
      expect(subject.host).to eq 'example.com'
    end
  end

  describe "#port" do
    it "returns the port" do
      expect(subject.port).to eq 443
    end
  end

  describe "#ssl?" do
    context "scheme is https" do
      it "returns true" do
        subject.scheme = 'https'
        expect(subject.ssl?).to eq true
      end
    end

    context "scheme is http" do
      it "returns false" do
        subject.scheme = 'http'
        expect(subject.ssl?).to eq false
      end
    end
  end

  describe "#http_connection" do
    it "behaves like HTTP client" do
      expect(subject.http_connection).to respond_to(:get)
      expect(subject.http_connection).to respond_to(:post)
      expect(subject.http_connection).to respond_to(:put)
      expect(subject.http_connection).to respond_to(:delete)
    end
  end

  describe "#absolute_url" do
    context "with no parameters" do
      it "returns a uri without path" do
        expect(subject.absolute_url).to eq "https://example.com"
      end
    end

    context "with parameters" do
      it "returns a uri with path" do
        expect(subject.absolute_url('/oauth/v2/authorize')).to eq "https://example.com/oauth/v2/authorize"
      end
    end
  end

  describe "#configure_ssl" do
  end

  describe "#redirect_limit_reached?" do
  end

  describe "#ssl_verify_mode" do
    context "ssl verify set to true" do
      it "returns OpenSSL::SSL::VERIFY_PEER" do
        subject.ssl = { :verify => true }
        expect(subject.send(:ssl_verify_mode)).to eq OpenSSL::SSL::VERIFY_PEER
      end
    end

    context "ssl verify set to false" do
      it "returns OpenSSL::SSL::VERIFY_NONE" do
        subject.ssl = { :verify => false }
        expect(subject.send(:ssl_verify_mode)).to eq OpenSSL::SSL::VERIFY_NONE
      end
    end
  end

  describe "ssl_cert_store" do
  end

  describe "#send_request" do
    before do
      @http_ok = OpenStruct.new(
        :code    => '200',
        :body    => 'success',
        :headers => {'Content-Type' => "application/json"}
      )
      @http_redirect = OpenStruct.new(
        :code    => '301',
        :body    => 'redirect',
        :headers => {'Location' => "https://example.com/members"}
      )
    end

    context "when method is not supported" do
      it "raises an error" do
        expect {subject.send_request(:patch, '/')}.to raise_error(SimpleHttp::Connection::UnhandledHTTPMethodError)
      end
    end

    context "when method is get" do
      it "returns an http response" do
        stub_get('/oauth/authorize').with(
          :query => {:client_id => '001337', :client_secret => 'abcxyz'},
          :headers => subject.default_headers
        )
        subject.send_request(:get, '/oauth/authorize', :params => {:client_id => '001337', :client_secret => 'abcxyz'})
      end
    end

    context "when method is delete" do
      it "returns an http response" do
        stub_delete('/users/1').with(
          :headers => subject.default_headers
        )
        subject.send_request(:delete, '/users/1')
      end
    end

    context "when method is post" do
      it "returns an http response" do
        stub_post('/users').with(
          :body => {:first_name => 'john', :last_name => 'smith'},
          :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
        )
        subject.send_request(:post, '/users', :params => {:first_name => 'john', :last_name => 'smith'})
      end
    end

    context "when method is put" do
      it "returns an http response" do
        stub_put('/users/1').with(
          :body => {:first_name => 'jane', :last_name => 'doe'},
          :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
        )
        subject.send_request(:put, '/users/1', :params => {:first_name => 'jane', :last_name => 'doe'})
      end
    end

    it "follows redirect" do

      stub_post('/users').with(
        :body => {:first_name => 'jane', :last_name => 'doe'},
        :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
      ).to_return(
        :status => 301,
        :body   => 'redirect',
        :headers => {'Location' => "https://example.com/members"}
      )

      stub_post('/members').with(
        :body => {:first_name => 'jane', :last_name => 'doe'},
        :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
      ).to_return(
        :status => 200,
        :body   => 'sucess'
      )

      subject.send_request(:post, '/users', :params => {:first_name => 'jane', :last_name => 'doe'})
    end

    it "respects the redirect limit " do
      subject.max_redirects = 1

      stub_post('/users').with(
        :body => {:first_name => 'jane', :last_name => 'doe'},
        :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
      ).to_return(
        :status => 301,
        :body   => 'redirect',
        :headers => {'Location' => "https://example.com/members"}
      )

      stub_post('/members').with(
        :body => {:first_name => 'jane', :last_name => 'doe'},
        :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
      ).to_return(
        :status => 301,
        :body   => 'redirect',
        :headers => {'Location' => "https://example.com/profiles"}
      )

      subject.send_request(:post, '/users', :params => {:first_name => 'jane', :last_name => 'doe'})
    end

    it "modifies http 303 redirect from POST to GET " do

      stub_post('/users').with(
        :body => {:first_name => 'jane', :last_name => 'doe'},
        :headers => {'Content-Type' => 'application/x-www-form-urlencoded' }.merge(subject.default_headers)
      ).to_return(
        :status => 303,
        :body   => 'redirect',
        :headers => {'Location' => "https://example.com/members"}
      )

      stub_get('/members').with(
        :headers => subject.default_headers
      ).to_return(
        :status => 200,
        :body   => ''
      )
      response = subject.send_request(:post, '/users', :params => {:first_name => 'jane', :last_name => 'doe'})

    end
  end
end

lib/simple_http/connection.rb
require 'zlib'
require 'addressable/uri'

begin
  require 'net/https'
rescue LoadError
  warn "Warning: no such file to load -- net/https. Make sure openssl is installed if you want ssl support"
  require 'net/http'
end

module SimpleHttp
  class Connection

    class UnhandledHTTPMethodError < StandardError; end
    class UnsupportedSchemeError < StandardError; end

    NET_HTTP_EXCEPTIONS = [
      EOFError,
      Errno::ECONNABORTED,
      Errno::ECONNREFUSED,
      Errno::ECONNRESET,
      Errno::EINVAL,
      Net::HTTPBadResponse,
      Net::HTTPHeaderSyntaxError,
      Net::ProtocolError,
      SocketError,
      Zlib::GzipFile::Error,
    ]

    attr_accessor :config, :scheme, :host, :port, :max_redirects, :ssl,
                  :user_agent, :accept, :max_redirects, :headers, :uri

    def self.default_options
      {
        :headers => {
          'Accept'     => 'application/json',
          'User-Agent' => "Simple HTTP gem #{SimpleHttp::Version}",
          'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3'

        },
        :ssl => {:verify => true},
        :max_redirects => 5
      }
    end

    def initialize(url, options={})
      @uri = Addressable::URI.parse(url)
      self.class.default_options.keys.each do |key|
        instance_variable_set(:"@#{key}", options.fetch(key, self.class.default_options[key]))
      end
    end

    def default_headers
      self.class.default_options[:headers]
    end

    def scheme=(scheme)
      unless ['http', 'https'].include? scheme
        raise UnsupportedSchemeError.new "#{scheme} is not supported, only http and https"
      end
      @scheme = scheme
    end

    def scheme
      @scheme ||= @uri.scheme
    end

    def host
      @host ||= @uri.host
    end

    def port
      _port = ssl? ? 443 : 80
      @port = @uri.port || _port
    end

    def absolute_url(path='')
      "#{scheme}://#{host}#{path}"
    end

    def ssl?
      scheme == "https" ? true : false
    end

    def ssl=(opts)
      raise "Expected Hash but got #{opts.class.name}" unless opts.is_a?(Hash)
      @ssl.merge!(opts)
    end

    def http_connection(opts={})
      _host   = opts[:host]   || host
      _port   = opts[:port]   || port
      _scheme = opts[:scheme] || scheme

      @http_client = Net::HTTP.new(_host, _port)

      configure_ssl(@http_client) if _scheme == 'https'

      @http_client
    end

    def send_request(method, path, opts={})
      headers         = @headers.merge(opts.fetch(:headers, {}))
      params          = opts[:params] || {}
      query           = Addressable::URI.form_encode(params)
      method          = method.to_sym
      normalized_path = query.empty? ? path : [path, query].join("?")
      client          = http_connection(opts.fetch(:connection_options, {}))

      if (method == :post || method == :put)
        headers['Content-Type'] ||= 'application/x-www-form-urlencoded'
      end

      case method
      when :get, :delete
        response = client.send(method, normalized_path, headers)
      when :post, :put
        response = client.send(method, path, query, headers)
      else
        raise UnhandledHTTPMethodError.new("Unsupported HTTP method, #{method}")
      end

      status = response.code.to_i

      case status
      when 301, 302, 303, 307
        unless redirect_limit_reached?
          if status == 303
            method = :get
            params = nil
            headers.delete('Content-Type')
          end
          redirect_uri = Addressable::URI.parse(response.header['Location'])
          conn = {
            :scheme => redirect_uri.scheme,
            :host   => redirect_uri.host,
            :port   => redirect_uri.port
          }
          return send_request(method, redirect_uri.path, :params => params, :headers => headers, :connection_options => conn)
        end
      when 100..599
        @redirect_count = 0
      else
        raise "Unhandled status code value of #{response.code}"
      end
      response
    rescue *NET_HTTP_EXCEPTIONS
      raise "Error::ConnectionFailed, #{$!}"
    end

  private

    def configure_ssl(http)
      http.use_ssl      = true
      http.verify_mode  = ssl_verify_mode
      http.cert_store   = ssl_cert_store

      http.cert         = ssl[:client_cert]  if ssl[:client_cert]
      http.key          = ssl[:client_key]   if ssl[:client_key]
      http.ca_file      = ssl[:ca_file]      if ssl[:ca_file]
      http.ca_path      = ssl[:ca_path]      if ssl[:ca_path]
      http.verify_depth = ssl[:verify_depth] if ssl[:verify_depth]
      http.ssl_version  = ssl[:version]      if ssl[:version]
    end

    def ssl_verify_mode
      if ssl.fetch(:verify, true)
          OpenSSL::SSL::VERIFY_PEER
      else
          OpenSSL::SSL::VERIFY_NONE
      end
    end

    def ssl_cert_store
      return ssl[:cert_store] if ssl[:cert_store]
      cert_store = OpenSSL::X509::Store.new
      cert_store.set_default_paths
      cert_store
    end

    def redirect_limit_reached?
      @redirect_count ||= 0
      @redirect_count += 1
      @redirect_count > @max_redirects
    end
  end
end